File: 2025/ts/src/day-3/part-1.ts
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';


const PATH = path.join(__dirname, '../../..', 'inputs/day-3/part-1/input.txt');

function maxPairD1(line:string): number {
  let firstMaxDigit = -1;
  let firstMaxDigitIndex = -1;

  for (let i = 0; i < line.length - 1; i++) {
    const num = parseInt(line[i], 10);
    if (num > firstMaxDigit) {
      firstMaxDigit = num;
      firstMaxDigitIndex = i;
    }
  }

  let secondMaxDigit = -1;

  for (let i = firstMaxDigitIndex + 1; i < line.length; i++) {
    const num = parseInt(line[i], 10);
    if (num > secondMaxDigit) {
      secondMaxDigit = num;
    }
  }

  return (firstMaxDigit * 10) + secondMaxDigit;
}

export async function day3Part1(): Promise<void> {
  const fileStream = fs.createReadStream(PATH);
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    try {
      let solution = 0;
      for await (const line of rl) {
        const maxPair = maxPairD1(line.trim());
        solution += maxPair;
      }
      console.log(`Solution: ${solution}`);
    } finally {
      rl.close();
    }
}

File: 2025/go/day-3/part-1.go
package day3

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	FILE_PATH = "../inputs/day-3/part-1/input.txt"
)

func parseDigit(b byte) (int, error) {
	return strconv.Atoi(string(b))
}

func maxPairD1(line string) int {
	firstMaxDigit := -1
	firstMaxDigitIndex := -1

	for i := 0; i < len(line)-1; i++ {
		num, err := parseDigit(line[i])
		if err != nil {
			fmt.Printf("Error parsing digit: %v\n", err)
			os.Exit(1)
		}
		if num > firstMaxDigit {
			firstMaxDigit = num
			firstMaxDigitIndex = i
		}
	}

	secondMaxDigit := -1

	for i := firstMaxDigitIndex + 1; i < len(line); i++ {
		num, err := parseDigit(line[i])
		if err != nil {
			fmt.Printf("Error parsing digit: %v\n", err)
			os.Exit(1)
		}
		if num > secondMaxDigit {
			secondMaxDigit = num
		}
	}

	return (firstMaxDigit * 10) + secondMaxDigit
}

func Day3Part1() {
	file, err := os.Open(FILE_PATH)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	pairSum := 0
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if err != nil {
			fmt.Printf("Error parsing line: %v\n", err)
			os.Exit(1)
		}

		maxPair := maxPairD1(line)
		pairSum += maxPair
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Solution: %d\n", pairSum)
}
